### ISSUE TITLE
lfs.concurrenttransfers does not speed up download of many small files

### ISSUE THREAD

**Issue by gebi** [git-lfs/git-lfs#1923]
Hi,

From searching the issue archive i found lfs.concurrenttransfers should speed up the transfer of many small files tracked with git-lfs.
I've a repo with about 9k small files tracked in git-lfs and both commands take about 2m30 to complete on a private server over direct LAN connection.

This two commands are equally slow to download
```
git -c lfs.concurrenttransfers=3  lfs clone
   vs.
git -c lfs.concurrenttransfers=50 lfs clone
```

The test repo https://github.com/larsxschneider/lfstest-manyfiles actually shows the same behaviour though, taking  about 3m20 to complete a clone here.
Setting lfs.batch to true does also not make any difference.

If those 2 commands are both equally slow then something else must limit the speed to get the 9k files transfered.
Befor starting to debug this and because i've read about some queue enhancments in git-lfs 2.0, does anyone have other suggestsions to speed this up?

Versions:
git-lfs/1.5.5 (GitHub; linux amd64; go 1.7.4; git c2dcd6f5)
git version 2.11.0
os: debian sid

**Comment by technoweenie** [comment]
All `lfs.concurrenttransfers` does is tells the transfer queue more requests in parallel. I don't know if `https://github.com/larsxschneider/lfstest-manyfiles` is a good repository to test this, since individual object downloads are so small. That said, I did personally see a big improvement increasing the transfers:

```
# lfs.concurrenttransfers=3
git lfs clone https://github.com/larsxschneider/lfstest-manyfiles  35.84s user 74.56s system 21% cpu 8:42.32 total

# lfs.concurrenttransfers=15
git lfs clone https://github.com/larsxschneider/lfstest-manyfiles   37.10s user 100.38s system 71% cpu 3:12.66 total

# lfs.concurrenttransfers=50
git lfs clone https://github.com/larsxschneider/lfstest-manyfiles   38.28s user 133.95s system 92% cpu 3:06.25 total
```

Maybe the problem is that Git LFS is not seeing the `-c lfs.concurrenttransfers=50` that you're passing to Git?

Also, I'm quite sure that there are lots of opportunities to optimize and improve LFS. I'd love more help in that area :)

**Comment by gebi** [comment]
Setting lfs.concurrenttransfers seems working because i can see the 50 open connections to the backend system ;).

@technoweenie we are basically seeing the same speed (3min) at the fastest settings, just that here even the default settings seem equally fast.

I'm asking because git-lfs is still WAY slower than it could/should be, maybe someone already has a clue on why it is so slow.

As the concurrency seems to not really matter it's maybe #1649 helps?

**Comment by technoweenie** [comment]
Possibly. I happened to be running on master with those changes in #1649. The main difference between the master branch and v1.5.5 is that v1.5.5 has to scan the 15k objects in `https://github.com/larsxschneider/lfstest-manyfiles` before it can even start downloading a single object. On v2.0-pre in master, it starts immediately. But that improvement isn't related to `lfs.concurrenttransfers`.

**Comment by gebi** [comment]
it seems i was totally wrong about it not being slower without lower concurrenttransfers settings, maybe i had it still configured in my home gitconfig.

Though the thing remains that it's painfull slow to download many small files with git-lfs.

**Comment by ttaylorr** [comment]
> Setting lfs.batch to true does also not make any difference.

Not related to this issue, but we should remove this setting before shipping v2.0.0 since we don't support the legacy API anymore. @technoweenie do you want to remove this or keep it (`lfs.batch` == `true` would imply a batch size of 1, or something)?

> Though the thing remains that it's painfull slow to download many small files with git-lfs.

There are definitely more things that I'd like to explore in continuing to make LFS faster. I'm curious what you mean by "small files", since it may not totally be in LFS's area of concern.

**Comment by ttaylorr** [comment]
Hi @gebi, I hope you don't mind the ~1-year later ping. I think that I am inclined to close this issue for now, since I can't determine a satisfiable trade-off between high-performance transfers for large objects and small objects at the same time.

Git LFS would like to batch many large objects so that it can avoid re-transferring objects that the server already has, but would like to _not_ batch small objects, so that it can transfer small amounts of data without having to first contact the server.

Right now, I am not sure what a long-term solution for this would be. I think that the best I can advise in the short-term is to take advantage of the flexibility of `.gitattributes` patterns to avoid tracking small files in Git. There are degenerate cases, of course, where you have small files that change _a lot_, and are thus not suited for Git, but rather Git LFS, which cannot accept them with great performance.

I think that there are large problems to tackle here, like binary diff-ing, protocol extensions and whatnot, but I believe that they are too large to discuss in this issue, and would benefit instead from a dedicated proposal.

I don't have time to explore this further at the current moment, but I would be more than happy to help answer questions or provide feedback on such a proposal. Until then, I am going to close this issue, but would be happy to re-open it at any point.

**Comment by mitar** [comment]
What we ended up doing is having a script which adds all files larger than 100 KB to LFS, and all other files to git, even if binary. Git is amazingly fast for small files, both transfer and checkout. And LFS is great for large files. And this solved our problem and we found the best balance. Our git repo is now 2 GB, and LFS data alltogether 70 GB. It is transferred at full link speed with minimal overhead.

**Comment by gebi** [comment]
@ttaylorr sorry i currently don't have the time to dig further into the low-leverl details of git-lfs, though from a high level point of view it is _FAR_ slower than what should be possible.

In the example below it's "only" 15k requests, that's what i find strange, on a local LAN as we have also tested that, that's possible within one second, not minutes.

Usually with the current speed of git-lfs it seems it does not properly parallelise the requests at all.

Short highlevel calculation:
- 500 parallel requests requested
- it takes 168s for only 15k requests

That means it would take 10 seconds per request if it could make the full 500 parallel requests (15000 / 168 / 500 = 0.1 req/s per "thread").
Something is fishy here i'd say, and that was the essence of my bug report.

```
% git -c lfs.concurrenttransfers=500 lfs clone https://github.com/larsxschneider/lfstest-manyfiles
Cloning into 'lfstest-manyfiles'...
remote: Counting objects: 15012, done.
remote: Total 15012 (delta 0), reused 0 (delta 0), pack-reused 15012
Receiving objects: 100% (15012/15012), 2.02 MiB | 1.42 MiB/s, done.
git -c lfs.concurrenttransfers=500 lfs clone   13.90s user 3.42s system 10% cpu 2:48.28 total

% git lfs version
git-lfs/2.4.0 (GitHub; linux amd64; go 1.8.3; git 1a98fa5c)
```

**Comment by gebi** [comment]
btw... one high level thing that also make me suspicious about the parallelising of git-lfs...

15k requests in 168seconds
that's about 90 req/s

that would mean an end-to-end latency of 11ms.
One request to github without any pipelineing takes about 700ms here, so we have a parallelisation of only 7, not 500 as requested.
Is there some throttling in the API?

I will retest it with our local git-lfs server here which can sustain way more than 20k req/s.

**Comment by ttaylorr** [comment]
> I will retest it with our local git-lfs server here which can sustain way more than 20k req/s.

Sounds good. Let me know.

**Comment by ttaylorr** [comment]
> @ttaylorr just re-tested with git-lfs 2.4 and git 2.7.4 and the performance problem still persists, though it seems it's not the download anymore but the actual checkout (even if `git lfs clone` is used, the actual checkout is on tmpfs and all machine cores are empty)

Hi @gebi, thanks for ensuring that you're on the latest version of Git LFS. Could you also upgrade your installation of Git? The latest is v2.17.0, and the earliest support for the improved clone time was released as a part of v2.11.0. To ensure that you're using this: (1) upgrade to the latest version of Git, and (2) run `git lfs install` to ensure that you've configured Git+Git LFS properly.
